File: app/src/main/java/com/example/soccernews/ui/NewsListAdapter.java
package com.example.soccernews.ui;

import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.soccernews.R;
import com.example.soccernews.data.model.News;
import com.example.soccernews.databinding.NewsItemBinding;
import com.squareup.picasso.Picasso;

public class NewsListAdapter extends ListAdapter<News, NewsListAdapter.ViewHolder>{

    public NewsListAdapter() {
        super(new DiffCallBack());
    }

    public FavoriteInterface favoriteInterface;

    @NonNull
    @Override
    public NewsListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        final NewsItemBinding binding = NewsItemBinding.inflate(inflater, parent, false);

        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull NewsListAdapter.ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    protected class ViewHolder extends RecyclerView.ViewHolder {
        private final NewsItemBinding binding;


        public ViewHolder(@NonNull NewsItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(News item){
            int color;

            binding.tvTitle.setText(item.getTitle());
            binding.tvBody.setText(item.getDescription());

            color = item.getFavorite() ? R.color.red : R.color.black;
            binding.ivFavorite.setColorFilter(
                    ContextCompat.getColor(binding.ivFavorite.getContext(), color));

            Picasso.get()
                    .load(item.getImage())
                    .fit()
                    .into(binding.ivThumbnail);

            binding.btnOpenLink.setOnClickListener(view -> {
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setData(Uri.parse(item.getLink()));
                itemView.getContext().startActivity(intent);
            });
            binding.ivShare.setOnClickListener(view -> {
                Intent intent = new Intent(Intent.ACTION_SEND);
                intent.setType("text/plain");
                intent.putExtra(Intent.EXTRA_TITLE, item.getTitle());
                intent.putExtra(Intent.EXTRA_TEXT, item.getLink());
                itemView.getContext().startActivity(Intent.createChooser(intent, "Share"));
            });
            binding.ivFavorite.setOnClickListener(view -> {
                item.setFavorite(!item.getFavorite());
                favoriteInterface.favoriteListener(item);
                notifyItemChanged(getAdapterPosition());
            });

        }

    }
}

class DiffCallBack extends DiffUtil.ItemCallback<News>{

    @Override
    public boolean areItemsTheSame(@NonNull News oldItem, @NonNull News newItem) {
        return oldItem == newItem;
    }

    @Override
    public boolean areContentsTheSame(@NonNull News oldItem, @NonNull News newItem) {
        return false;
    }
}

File: app/src/main/java/viana/alison/soccernews/ui/adapters/NewsListAdapter.kt
package viana.alison.soccernews.ui.adapters

import android.content.Intent
import android.net.Uri
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import viana.alison.soccernews.data.model.News
import com.squareup.picasso.Picasso
import viana.alison.soccernews.R
import viana.alison.soccernews.databinding.ItemNewsBinding

class NewsListAdapter : ListAdapter<News, NewsListAdapter.ViewHolder>(DiffCallBack()) {

    var favoriteButtonListener: (News) -> Unit = {}

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsListAdapter.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ItemNewsBinding.inflate(inflater, parent, false)

        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(
        private val binding: ItemNewsBinding
        ): RecyclerView.ViewHolder(binding.root) {

        fun bind(item: News) {

            binding.tvTitle.text = item.title
            binding.tvBody.text = item.description

            val color = if (item.favorite) R.color.red else R.color.black
            binding.ivFavorite.setColorFilter(
                ContextCompat.getColor(binding.ivFavorite.context, color))

            Picasso.get()
                .load(item.image)
                .fit()
                .into(binding.ivThumbnail)

            binding.btnOpenLink.setOnClickListener {
                Intent(Intent.ACTION_VIEW).apply {
                    data = Uri.parse(item.link)
                    itemView.context.startActivity(this)
                }
            }
            binding.ivShare.setOnClickListener {
                Intent(Intent.ACTION_SEND).apply {
                    type = "text/plain"
                    putExtra(Intent.EXTRA_TITLE, item.title)
                    putExtra(Intent.EXTRA_TEXT, item.link)
                    itemView.context.startActivity(Intent.createChooser(this, "Share"))
                }
            }
            binding.ivFavorite.setOnClickListener {
                item.favorite = !item.favorite
                favoriteButtonListener(item)
                notifyItemChanged(adapterPosition)
            }

        }
    }

}

class DiffCallBack : DiffUtil.ItemCallback<News>() {
    override fun areItemsTheSame(oldItem: News, newItem: News) = oldItem == newItem
    override fun areContentsTheSame(oldItem: News, newItem: News) = oldItem.id == newItem.id
}